javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,223 @@
+
+### Use cases
+
+What can we do with those hashes, You have probably seen it in school where a folder is organized into smaller folders, each folder holds information about people with the same first letter, so we get something like this:
+
+```
+C:
+  carlos.txt
+  carla.txt
+B:
+  bryan.txt
+  bob.txt
+  beatrice.txt
+  bella.txt
+  benjamin.txt
+  bianca.txt
+```
+
+If we get new student in our school with the name `"Carlos"` we can run our hash function to find out what folder do we assign them in `hash("Carlos") -> "C"` so we put `"Carlos"` in the directory labeled `C`
+
+You might have spotted a problem, what if our school is populated with many people that share the same first letter `C`? Then we will have a directory with labeled `C` that have too many names that start with `"C"` and the other directories are empty. We need to rework our hash function, to eliminate more duplication and separate our students, while keeping constant structure in folders.
+
+```javascript
+function hash(name, surname) {
+  return name.charAt(0) + surname.charAt(0);
+}
+```
+
+Instead of just taking the first name letter, we take the first name and last name letters. `"Carlos Smith"` will have a hash code of `"CS"` this will surely spread our students in more directories, it will eliminate many duplicate hash codes from being generated.
+
+![Example of hashing a name using first name's first letter, and last name's first letter](./project_hash_map/imgs/00.png)
+
+But it still doesn't solve our problem, what if we have a very popular first name's first letter and surname's first letter? Then we are again with directories that are empty and other directories that are full, we need it easier for us to find the person we are looking for, so let's rework our hash code.
+
+```javascript
+function stringToNumber(string) {
+  let hashCode = 0;
+  for (let i = 0; i < string.length; i++) {
+    hashCode += string.charCodeAt(i);
+  }
+
+  return hashCode;
+}
+
+function hash(name, surname) {
+  return stringToNumber(name) + stringToNumber(surname);
+}
+```
+
+Using this technique we don't only take the first letters into consideration, now we take the whole name and convert it into numbers.
+
+You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true, this would make it unique for each name, but there are other reasons that we have chosen a number:
```

Comment:
```suggestion
You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true. This would make it unique for each name, but there are some good reasons to choose a number:
```